docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+
+---
+
+### General Design
+
+![](../img/ALCProxyDesign.svg)
+
+The ProxyBuilder API will generate five separate objects:
+* `ProxyBuilder`: Responsible for creating `ProxyObject`, `ClientObject`, and `DispatchBuilder` and returns the instance of the desired proxy to the user for use.
+* `DispatchBuilder`: Constructs `ServerObject` in the target ALC.
+* `ProxyObject`: The object that the user has access to, it allows for method calls to run on the `TargetObject` in the target ALC. When a method is called, `ProxyObject` pipes its command to the `ServerObject`, which will eventually from the user context to the new proxy now in the target ALC using binary serialization.
+* `ClientObject`: Sends messages to the `ServerObject` in the target ALC requesting `TargetObject`'s method to be called.
+* `ServerObject`: Invokes `TargetObject`'s methods, and pipes the output back to `ClientObject` and the user ALC.
+
+When a method is called, `ProxyObject` pipes its command to the `ClientObject`, which will serialize or encode the method information and any parameters passed to the `ProxyObject`, and send them to `ServerObject` through some sort of request. `ServerObject` then deserializes the data, and uses reflection to invoke the method in `TargetObject`. Once the method runs and `ServerObject` gets the returned object, it sends it back to `ClientObject`, which sends it back to `ProxyObject` to be "returned" to the user.
+
+A `DispatchProxy` can be used to create a "fake" object of a given type, which acts as if it was the instance we wanted to use. Having `ProxyObject` inherit from `DispatchProxy` allows us to easily add more extra actions when we call the object, such as deserialization of objects across the type barrier (See the "Type-Isolation Barrier" section below).
+
+This design is dependent on the ability for the `ProxyBuilder` to create dynamic assemblies and types within a different ALC/Process. Currently we may be able to do so by loading the `DispatchBuilder` type into the ALC directly, and then creating the proxy object from there, while also using a similar builder to create the user-side proxy that make gRPC calls. We can also wait (See [#38426](https://github.com/dotnet/corefx/issues/38426 )), but this option will not be available until said issue can be fixed, or a comparable fix is made.
+
+### Variant for ALC Revisions with `DispatchProxy`
+
+![](../img/ALCProxyAlternate.svg)
+
+One thing that's nice about `DispatchProxy` is that it allows for us to load a target Type directly from the target ALC to the user ALC, allowing us to remove the `ServerObject`, and instead just call the object's methods directly. This only works for in-process communications, though, and by removing `ServerObject`, we may lose the ability to allow for consistent unloadability (due to how the `TargetObject` may return objects), so this approach needs to be investigated further.
```

Comment:
Consistent unloadability is not something we want to support.
It is not an option for the target of a proxy to be unloaded while the proxy is held.  This feature was deliberately removed from `AppDomain`s as it introduced far too many bugs. It also made it extremely difficult to use the proxies.

Reply:
@sdmaclea I would love to see data for this. The problem without proxies is that we have only a rather poor story for making unloadability even remotely predictable. I do agree that proxies alone don't solve the problem 100%, but they can go a long way to do so. My understanding is that we deliberately removed the feature which reliably stops execution in a specific `AppDomain` - the "free resources" part of `AppDomain` unload is still very valuable I think.